File: python/pkt_listen.py
import wtools
import socket

def listen(addr :str, port :int, path :str):
    listen_fd = socket.socket()
    listen_fd.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    listen_fd.bind((addr, port))
    listen_fd.listen(5)
    
    print(f'waiting client connected...')
    client_fd, client_addr = listen_fd.accept()
    
    print(f'client connected: [{client_addr[0]}:{client_addr[1]}]')
    pkt = wtools.packet()

    print('recv...')
    content = pkt.recv(client_fd)
    file_data = content[3]
    print('recv done.')

    print('send.')
    pkt.send(client_fd, b'done.')

    print(f'save to {path}')
    with open(path, 'wb') as f:
        f.write(file_data)

    client_fd.close()
    listen_fd.close()

if __name__ == '__main__':
    listen('0.0.0.0', 9981, 'test_file')

File: test/pkt_listen.cpp
#include <network/socket.hpp>

#include <iostream>
#include <fstream>

#include <network/socket.cpp>
#include <network/exception.cpp>

using namespace std;

namespace wmkc_test {
    class PacketBuffer {
    public:
        wByte *buffer;
        wU32 length;

        PacketBuffer() {}
    };

    void listen(string listen_addr, uint16_t listen_port) {
        wmkc::net::Socket listen_fd(AF_INET, SOCK_STREAM, IPPROTO_TCP);
        int enable = 1;
        setsockopt(listen_fd.fd, SOL_SOCKET, SO_REUSEADDR, (char *)&enable, sizeof(enable));
        listen_fd.bind(listen_addr, listen_port);
        listen_fd.listen(5);

        printf("waiting client connected...\n");
        auto client_fd = listen_fd.accept();
        printf("client connected.\n");

        string string_buffer = client_fd.recv(4);
        string buffer{};
        wU32 length{};

        memcpy(&length, string_buffer.data(), 4);

        printf("start recv...\n");
        while(length) {
            string temp = client_fd.recv(wmkc_min(2048, length));
            if(temp.empty()) {
                break;
            }
            buffer.append(temp);
            length -= temp.size();
        }

        printf("file writing...\n");
        fstream f("C:/Users/sn/Desktop/f_tp_test/test", ios::out | ios::binary);
        f.write(buffer.c_str(), buffer.size());
        f.close();

        client_fd.send("done.");

        client_fd.close();
        listen_fd.close();
    }
}

int main()
{
    WSADATA ws;
    WSAStartup(MAKEWORD(2,2), &ws);

    wmkc_test::listen("0.0.0.0", 80);

    WSACleanup();
}
